### PR TITLE
Fix boost::graph_traits issues 

### LINKED ISSUE #2570: Missing types in BGL concepts
`HalfedgeGraph`, `FaceGraph`, `HalfedgeListGraph`, and `FaceListGraph` concepts do not document the types the user must provide through `graph_traits` .

### LINKED ISSUE #2629: Things that are "documented" by error in Surface_mesh
In the [documentation of `Surface_mesh` in CGAL-4.11](https://doc.cgal.org/4.11/Surface_mesh/annotated.html), one can see documentation pages for specializations of `graph_has_property` and `property_map`. That is probably not wanted.

image

### PR DESCRIPTION
**Pull request by afabri**
## Summary of Changes

- `null_halfedge` is a static function in the scope of `boost::graph_traits`.
-  Document that `boost::graph_traits<HalfedgeGraph>::halfedge_descriptor` is an associated type.
- The same for face, and for halfedge and face iterator.
## Release Management

* Affected package(s): BGL
* Issue(s) solved (if any): fix #2570, fix #2572, fix #2629

### CODE REVIEW

File: BGL/doc/BGL/Concepts/FaceGraph.h [hunk 1]
```diff
@@ -8,13 +8,23 @@ maintain faces described by halfedges, to provide access from a face to
 an incident halfedge, and to provide access from a halfedge to its incident
 face. 
 
+<h4>Associated Types:</h4>
+
+`boost::graph_traits<FaceGraph>::face_descriptor`
+A face descriptor corresponds to a unique face in an abstract graph instance.
+A face descriptor must be `DefaultConstructible`, `Assignable`, and `EqualityComparable`.
```
Comment (sloriot) [inline BGL/doc/BGL/Concepts/FaceGraph.h]: Some algorithm requires descriptors to be hashable. We should probably add the requirement here.

Comment (afabri) [reply BGL/doc/BGL/Concepts/FaceGraph.h]: How do you see that for vertex_descriptor. It is documented in the base concept in boost, and does not mention hashable. Shall we just repeat the vertex type in our documentation.

Comment (sloriot) [reply BGL/doc/BGL/Concepts/FaceGraph.h]: Either we request it in functions that use it or we add it in `HalfedgeGraph` as you suggested.

**Comment by lrineau** [comment]
Doc errors:
https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-4.12-Ic-91/BGL.log

**Comment by lrineau** [comment]
(I mark PR with the label `not yet approved`, to remember that the PR is still under development.)

**Comment by lrineau** [comment]
Do you find the result easy to read? For example:

https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-4.12-Ic-104/master/BGL/classFaceGraph.html

The name of the type and its documentation are in the same line, without any separation.
- Should not it be implemented with things like `<dl>`, `<dd>, `<dt>`, or with a `<table>`?
- Is that the first time we document associated types?

**Comment by afabri** [comment]
I guess one could use a macro similar as we use it for named parameters.

**Comment by afabri** [comment]
[Associated Types](https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-4.12-Ic-105/output/BGL/classHalfedgeGraph.html) now look better.

**Comment by lrineau** [comment]
This PR conflicts with `master`:
> *Conflicting files*
> `Surface_mesh/include/CGAL/boost/graph/graph_traits_Surface_mesh.h`
> `Surface_mesh/include/CGAL/boost/graph/properties_Surface_mesh.h`

I suggest a rebase.

**Comment by lrineau** [comment]
Actually the conflicts are simple. They can be resolved with the Github web interface (button "Resolve conflicts" below).

**Comment by afabri** [comment]
done

**Pull request merged**